### PR TITLE
feat(sdk): Validating and Testing Fern Definitions

### PR DESCRIPTION
**Pull request by jonahkaye**
Refs: #1212, #1219, #1371

### Description

- fixing the entire fern definition to make it actually work
- writing testing for all three sdks. 
- moving testers folders into their respective subdirectories in sdk
- one or two misc docx bugs fixed. 
- fern definitions for webhook types too
- related to but not dependent on https://github.com/metriport/metriport-java/pull/2 
- merged this PR into mine: https://github.com/metriport/metriport/pull/1188/
- merged this PR into mine: https://github.com/metriport/metriport/pull/1220/
- merged this PR into mine: https://github.com/metriport/metriport/pull/1326
- merged this PR into mine: https://github.com/metriport/metriport/pull/1359

### Todo

- [x] review fern definitions and fix nits. 
- [x] fern generator larger changes: FHIR reference and document upload
- [x] finish writing tests for python and java
- [x] Update docx 
- [ ] integrate fern sdk testing into automatic testing in ci/cd? 

### Testing

- This PR is 99% testing! I put the local testing of fern in each of the respective python, java, typescript folders under in the folder tester-local. Each tester-local roughly has the same tests.
- The not local testing of each sdk is more sporadic, which makes sense since the sdks are not deployed. We might want to add an item to the go live to fully test the deployed sdks before sending sdk to customers. 
- The testing framework for java is a little bit funky. The local testing uses gradle but the testing for the actual sdk uses maven. I used what deep provided since I am really not so familiar with all the piping around running java code. 
- The not local testing for typescript still uses our current ts-sdk. It makes sense to rewrite this once we have the new ts-sdk deployed. 

### Note:
- If when you look at the tester-local files, you think something should be renamed or a type looks funny, all that is of course changeable
- You can build fern and run all local tests by running `npm run test-fern-local` from oss root

### Release Plan

- [x] test
- [ ] merge
- [ ] proceed with the tasks in go live ticket, which is source of truth for all things Fern rn

### CODE REVIEW

File: fern/definition/medical/patient.yml
Comment (jonahkaye) [inline fern/definition/medical/patient.yml:1]: changes here are most important. There were a few places the definitions here were diverging from the api specs

File: fern/definition/medical/patient.yml
Comment (dsinghvi) [inline fern/definition/medical/patient.yml:192]: @jonahkaye can you try `optional<list<Contact>>` and lets delete the `Contacts` type?

File: docs/medical-api/api-reference/patient/delete-patient.mdx
Comment (jonahkaye) [inline docs/medical-api/api-reference/patient/delete-patient.mdx:9]: modified to match endpoint definition. Question of why facility ID should be provided at all here? Like if its not nessecary its kinda just confusing and superfluous no?

File: fern/definition/medical/document.yml
Comment (jonahkaye) [inline fern/definition/medical/document.yml:144]: so this requestId in the docs doesnt have a required flag next to it. But ofc its always returned by our endpoint. We arent really consistent about that. Some endpoints have required in the response next to values that are always returned and others dont. Maybe not a big deal but wanted to note

**Comment by dsinghvi** [comment]
@jonahkaye Filed an issue [here](https://github.com/fern-api/fern-python/issues/426) to track the failing python generator. 

Beyond that, the reason `fern check` is failing is because some of the examples are not correct and you should be able to resolve based on the error messages.

File: fern/definition/medical/document.yml
Comment (jonahkaye) [inline fern/definition/medical/document.yml:219]: this is temporary until I know how to import the actual fern fhir reference spec

File: fern/definition/medical/document.yml
Comment (jonahkaye) [inline fern/definition/medical/document.yml:289]: We only use this schema in the return for DocumentQueries in the webhook. Unless we are going ti implement any validation of webhook responses (I don't think thats part of the sdk as of now but maybe it should be?) then this should all be deleted

File: fern/definition/medical/document.yml
Comment (jonahkaye) [inline fern/definition/medical/document.yml:369]: Internally we require that uploaded document references have a description, type, and context, not to the exclusion of other resource types. We should be able to use Fern import types here and then ssimplify this greatly too. 
We also need to be able to capture the logic that we require certain properties but that other remaining ones are optional and permitted.

File: fern/definition/medical/document.yml
Comment (jonahkaye) [inline fern/definition/medical/document.yml:140]: I followed the docx, but maybe we should decide to either call this createDocumentReference or uploadDocument and then stick with that for the entirety of the defintion instead of having it half and half

File: fern/definition/medical/fhir.yml
Comment (jonahkaye) [inline fern/definition/medical/fhir.yml:9]: temporarily commented out until Fern resolves bug on their end: https://github.com/fern-api/fern-python/issues/426

File: fern/definition/medical/fhir.yml
Comment (jonahkaye) [inline fern/definition/medical/fhir.yml:181]: Refactor with FHiR fern imports

File: fern/definition/medical/fhir.yml
Comment (jonahkaye) [inline fern/definition/medical/fhir.yml:76]: added missing endpoint

File: fern/definition/medical/document.yml
Comment (jonahkaye) [reply fern/definition/medical/document.yml:135]: This was a missing endpoint I added

File: fern/definition/commons.yml
Comment (jonahkaye) [inline fern/definition/commons.yml:88]: refactor. removable with Fhir spec import

File: fern/definition/medical/fhir.yml
Comment (jonahkaye) [inline fern/definition/medical/fhir.yml:92]: waiting on deep to get back to me on the fix here since this example doesnt compile

**Review by jonahkaye — COMMENTED** [review]
Reviewed the PR

File: fern/definition/medical/fhir.yml
Comment (jonahkaye) [inline fern/definition/medical/fhir.yml:102]: fix coming for this from fern. resource shouldnt need to be empty here in order to build

File: fern/definition/medical/webhooks.yml
Comment (jonahkaye) [inline fern/definition/medical/webhooks.yml:123]: I generally just kept the types names as the same ones in the App. Idk if we want to change that

**Review by jonahkaye — COMMENTED** [review]
Added webhook types

File: fern/definition/medical/document.yml
Comment (Goncharo) [inline fern/definition/medical/document.yml:98]: I know this is part of the docs, but the description doesn't make sense - we're not only searching the contents of the DocRef, but the contents of the document itself as well. Let's update the description accordingly in both places.

File: fern/definition/medical/fhir.yml
Comment (Goncharo) [inline fern/definition/medical/fhir.yml:42]: nit: is there a way to specify formatting here? ie be able to highlight `html` and `pdf`

File: fern/definition/medical/fhir.yml
Comment (Goncharo) [inline fern/definition/medical/fhir.yml:105]: nit: we don't actually ever parse to `Appointment` resources - perhaps we can make all the examples in the docs something more generic, like `Observation`?

File: fern/definition/medical/fhir.yml
Comment (Goncharo) [inline fern/definition/medical/fhir.yml:223]: ditto - suggest not using `Appointment` as an example

File: fern/definition/medical/patient.yml
Comment (Goncharo) [inline fern/definition/medical/patient.yml:125]: ditto - double check the formatting of `html` and `pdf`

**Review by Goncharo — APPROVED** [review]
generally LGTM - conditional approval on the comments getting addressed

File: fern/definition/medical/document.yml
Comment (jonahkaye) [reply fern/definition/medical/document.yml:98]: Replacing with:

"Value to search within the document reference and the actual contents of the document (minimum 3 chars)."

File: fern/definition/medical/fhir.yml
Comment (jonahkaye) [reply fern/definition/medical/fhir.yml:42]: yah just wrap in backticks

**Pull request merged**
